sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/QueryPlanConstraints.scala

```diff
@@ -33,7 +33,8 @@ trait QueryPlanConstraints extends ConstraintHelper { self: LogicalPlan =>
         .union(inferAdditionalConstraints(validConstraints))
         .union(constructIsNotNullConstraints(validConstraints, output))
         .filter { c =>
-          c.references.nonEmpty && c.references.subsetOf(outputSet) && c.deterministic
+          (c.references.nonEmpty && c.references.subsetOf(outputSet) || c.references.isEmpty) &&
```

Comment:
Hm, it's a bit weird that we consider the constraints are valid even for the constraint that holds the reference not being referred. Shouldn't we handle it in `InferFiltersFromGenerate` rule itself?

Reply:
cc @tanelk and @maropu (from https://github.com/apache/spark/commit/17eebd72097ee65e22cdaddf375e868074251f5a)

Reply:
I agree with @HyukjinKwon . In the `InferFiltersFromGenerate` we ignore expressions `e.children.forall(_.foldable)`. Unfortunatelly the `ScalaUDF` is never foldable (perhaps it could be). I would recomend turning this optimizer of, when the child expression is instance of `org.apache.spark.sql.catalyst.expressions.UserDefinedExpression`.